Stakers can withdraw airdropped tokens.

sausage-dog opened the pull request:
## Description

Requesting to make a pull request to **gem-farm** repo in order to add additional functionality regarding BONK and other airdropped tokens in the vault ATA. Instruction added to **gem-bank** in file [clean_vault.rs](https://github.com/sausage-dog/gem-farm/blob/main/programs/gem_bank/src/instructions/clean_vault.rs) allows to withdraw the SPL token balance to the account provided.

## Current Behaviour 

Users are unable to withdraw the airdropped tokens currently stuck in the vault PDA managed by them. 

## New Behaviour

Depositor able to withdraw any  SPL token held in an ATA managed by their vault PDA to any address. 


## Motivation and Context

Lot's of people have tokens that they are unable to claim as the withdraw authority is the program and it does not have the correct logic to allow them to withdrawal airdropped tokens currently stuck.

## How Has This Been Tested?


Tests are added onto the current gem-farm test script, showing seamless integration.
There are 3 tests included in file [gem-clean-vault.test.ts](https://github.com/sausage-dog/gem-farm/blob/main/tests/gem-bank/gem-clean-vault.test.ts).


For local testing: clone the repo - yarn - anchor test - (swap out programs IDs). 
Note: works with ts-mocha@10.0.0 due to Anchor.

No changes to the **gem-farm** program.

ilmoi commented on tests/gem-bank/gem-clean-vault.test.ts:
```diff
@@ -0,0 +1,528 @@
+      // bonk set-up
+      bonkToken = await Token.createMint(
+        _provider.connection,
+        PaYeR,
```
nit: is there a reason for this weird camelcasing?

sausage-dog replied:
Easier to search. Changed to txPayer.

ilmoi commented on tests/gem-bank/gem-clean-vault.test.ts:
```diff
@@ -0,0 +1,528 @@
+        recipient.publicKey
+      );
+      // Derive ATA of the bonk for recipient
+      [vaultAta_bonk] = await PublicKey.findProgramAddress(
```
is there a reason the same ix above doesn't work to create a program-owned TA?

sausage-dog replied:
Need to manually assemble a transaction to pay for the ATA creation. This `getOrCreateAssociatedAccountInfo(` only works for keypair accounts in this version. With newer solana/spl-token package version it does it for you in the background (solana/spl-token > 2.0).

ilmoi commented on tests/gem-bank/gem-clean-vault.test.ts:
```diff
@@ -0,0 +1,528 @@
+      // No change to fake bonk balance for the vault
+      assert.equal(vaultFakeBonkEnd, vaultFakeBonkStart);
+      // Transfered all the bonk from the vault to the recipient
+      assert.equal(vaultBonkStart, recipientBonkEnd - recipientBonkStart);
```
🔥

ilmoi commented on tests/gem-bank/gem-clean-vault.test.ts:
```diff
@@ -0,0 +1,528 @@
+      assert.equal(vaultBonkStart, recipientBonkEnd - recipientBonkStart);
+    });
+
+    it('Fake bonk SPL withdrawl after airdrop (caller has deposit)', async () => {
```
am I missing something or is this test entirely redundant? like it has 2 tokens just like the previous one, you withdraw one, keep the other. Maybe missing something

sausage-dog replied:
Redundant, just sanity check.

ilmoi replied:
if it's redundant then why keep it?

less code = better

sausage-dog commented on programs/gem_bank/src/instructions/clean_vault.rs:
```diff
@@ -0,0 +1,56 @@
+    let creator: &Pubkey = &&vault.creator;
+
+    // Seeds
+    let seed = &[b"vault".as_ref(), bank.as_ref(), &&vault.creator.as_ref()];
```
We would use `vault.vault_seeds()` if we were withdrawing from the gem box, but we are withdrawing from the vault itself which has a different PDA scheme. Are we correct to think that vault derived ATA owns the SPL so we sign with vault, rather than gem box?

ilmoi commented on programs/gem_bank/src/instructions/withdraw_tokens_vault.rs:
```diff
@@ -0,0 +1,78 @@
+
+    let (gem_box_pda, _) = anchor_lang::prelude::Pubkey::find_program_address(
+        &[
+            b"gem_deposit_receipt".as_ref(),
```
wrong seeds, this is gem_deposit_receipt, not gem_box

ilmoi commented on tests/gem-bank/gem-clean-vault.test.ts:
```diff
@@ -0,0 +1,528 @@
+  });
+}
+
+describe('gem clean vault', () => {
```
test files still use "clean vault" everywhere?

ilmoi commented on tests/gem-bank/gem-clean-vault.test.ts:
```diff
@@ -0,0 +1,528 @@
+      );
+    });
+  });
+});
```
where is the test that proves you can't withdraw a gem-box token using this ix, as we discussed in tg?

sausage-dog commented:
## Changes to original PR

After reading your initial comments we fixed issues raised, but missed the new test file when commiting, so you might have been reviewing an old test file. New test file is [withdraw-token-vault.test.ts](https://github.com/gemworks/gem-farm/blob/c3f6e5aac2655fadf6bb24de710038f580f9a649/tests/gem-bank/withdraw-token-vault.test.ts)




- [This redundant test is now removed from the tests](https://github.com/gemworks/gem-farm/pull/121#discussion_r1066271917)
- We used the seed from this [comment](https://github.com/gemworks/gem-farm/pull/121#discussion_r1065960101) which had `gem_deposit_receipt` rather than the `gem_box` in the seed derivation
- Added a [test](https://github.com/gemworks/gem-farm/blob/c3f6e5aac2655fadf6bb24de710038f580f9a649/tests/gem-bank/withdraw-token-vault.test.ts#L361) that show innability to withdraw from the [gembox](https://github.com/gemworks/gem-farm/pull/121#discussion_r1066274278)
- For this [comment](https://github.com/gemworks/gem-farm/pull/121#discussion_r1066271160) we changed the test name to reflect the name of the instruction being tested.
- For this [comment](https://github.com/gemworks/gem-farm/pull/121#discussion_r1066273689) changed everywhere function names to be for `token` and not `tokens`.
- It is not possible to sign for the vault with `.with_signer(&[&vault.vault_seeds()]),` as suggested [here](https://github.com/gemworks/gem-farm/pull/121#discussion_r1066269991). Seeds for the vault and gem are different.


This is the scheme for vault:

            b"vault".as_ref(),
            bank.key().as_ref(),
            creator.key().as_ref(),
        
        
This is the scheme for gem box:

            b"gem_box".as_ref(),
            vault.key().as_ref(),
            gem_mint.key().as_ref(),

`vault_seeds` returns seed for the authority account, but this is not the one we use as we are signing for the vault.

ilmoi commented:
closing in favor of https://github.com/gemworks/gem-farm/pull/123 as discussed in telegram
